File: packages/fluxguard-java/src/main/java/com/fluxguard/store/LocalStore.java
package com.fluxguard.store;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public final class LocalStore implements Store {
  private final Map<String, Entry> data = new ConcurrentHashMap<>();
  private final java.util.function.LongSupplier clock;

  public LocalStore() {
    this(System::currentTimeMillis);
  }

  public LocalStore(java.util.function.LongSupplier clock) {
    this.clock = clock;
  }

  private record Entry(String value, long expiresAt) {}

  @Override
  public String get(String key) {
    Entry e = data.get(key);
    long now = clock.getAsLong();
    if (e == null) {
      return null;
    }
    if (e.expiresAt > 0 && now > e.expiresAt) {
      data.remove(key);
      return null;
    }
    return e.value;
  }

  @Override
  public void set(String key, String value, long ttlMs) {
    long expiresAt = ttlMs > 0 ? clock.getAsLong() + ttlMs : 0;
    data.put(key, new Entry(value, expiresAt));
  }

  @Override
  public void del(String key) {
    data.remove(key);
  }

  @Override
  public List<Object> evalScript(String script, List<String> keys, List<String> args) {
    throw new UnsupportedOperationException("LocalStore does not support evalScript");
  }

  public void clear() {
    data.clear();
  }
}

File: packages/fluxguard-node/src/store/LocalStore.ts
import type { Store } from "./Store.js";

type Entry = { value: string; expiresAt: number };

export class LocalStore implements Store {
  private readonly data = new Map<string, Entry>();

  constructor(private readonly clock: () => number = () => Date.now()) {}

  async get(key: string): Promise<string | null> {
    const e = this.data.get(key);
    if (!e) return null;
    const now = this.clock();
    if (e.expiresAt > 0 && now > e.expiresAt) {
      this.data.delete(key);
      return null;
    }
    return e.value;
  }

  async set(key: string, value: string, ttlMs: number): Promise<void> {
    const expiresAt = ttlMs > 0 ? this.clock() + ttlMs : 0;
    this.data.set(key, { value, expiresAt });
  }

  async del(key: string): Promise<void> {
    this.data.delete(key);
  }

  async evalScript(): Promise<unknown[]> {
    throw new Error("LocalStore does not support evalScript — use algorithm local path");
  }

  /** Test helper: clear all keys */
  _clear(): void {
    this.data.clear();
  }
}
